### ISSUE TITLE
Can I use Arryalist as a playlist with concatenatingmediasource?

### ISSUE THREAD

**Issue by isuru18113** [google/ExoPlayer#7828]
Recently I am using Exoplayer for my new music player and Exoplayer is a great media player ever. I have build half of my app and now I'm trying to let the users play the next song or previous song. All the music data have stored in firebase and I am streaming mp3. Now I want to play the playlist which I stored as ArrayList in my database, Here is my ArrayList structure,
 **playlist**---   _arrray_
 

- **[0]**   _map_
**musicurl:  0mp3 link**
**musicimage: 0image link**
**musicname: 0helloworld**

- **[1]**   _map_
**musicurl:  1mp3 link**
**musicimage: 1image link**
**musicname: 1helloworld**

- **[...]**   _map_

I could play the playlist like this, 
``` 
public static final Sample[] SAMPLES = new Sample[] {
      new Sample(
          "https://storage.googleapis.com/automotive-media/Jazz_In_Paris.mp3",
          "audio_1",
          "Jazz in Paris",
          "Jazz for the masses",
          R.drawable.album_art_1),`

` new Sample(
          "https://storage.googleapis.com/automotive-media/The_Messenger.mp3",
          "audio_2",
          "The messenger",
          "Hipster guide to London",
          R.drawable.album_art_2)};
```
[https://github.com/google/ExoPlayer/blob/io18/audio-app/src/main/java/com/google/android/exoplayer/audiodemo/Samples.java](url)

Adding one by one like above is totally fine for few songs, But is there any possible way to achieve my goal? or is there any different way to stream mp3s as a playlist from the database?
Thank you!

**Comment by marcbaechinger** [comment]
If you construct the `ConcatenatingMediaSource` by passing the `lazyPreparation` flag set to true, you don't need to care much about the size of the playlist. The media sources will be prepared on demand only and does not cause a big overhead when the player is initially prepared.

```
    List<Map> items = new ArrayList<>();
    ProgressiveMediaSource.Factory factory = new ProgressiveMediaSource.Factory(dataSourceFactory);
    ConcatenatingMediaSource concatenatingMediaSource = new ConcatenatingMediaSource(
        /* isAtomic= */ false,
        /* useLazyPreparation= */ true,
        new ShuffleOrder.DefaultShuffleOrder(/* length= */ 0));
    for (int i = 0; i < items.size(); i++) {
      Uri uri = Uri.parse(items.get(i).get("musicurl"));
      concatenatingMediaSource.addMediaSource(factory.createMediaSource(uri));
    }
    player.prepare(concatenatingMediaSource);
```

If you set the `lazyPreparation` flag to true, adding all the sources is almost the same like adding them to a simple list like an `ArrayList`. No further action is taken by the player in lazy preparation mode for other than the media source currently being played.

> Adding one by one like above is totally fine for few songs, But is there any possible way to achieve my goal?

There is no direct support for iterating over a cursor of a database or a realtime database like Firebase, if that's what you mean. If your goal is different then just adding the items in the list to the player's playlist, please add some more details about your goal.

**Comment by isuru18113** [comment]
Thank you very much for your quick response and your answer also very clear. My goal is to play an ArrayList which I stored in firestore. I cannot add one by one to my java class like above the SIMPLES example. What I'm trying is,
The user selects the artist -----> send to the song list(PLAYLIST)---------->When user click a song in the playlist it should be played and if he wants to skip the song ` android:id="@+id/exo_next"` or listen to the previous song `android:id="@+id/exo_prev"`
using default ids. When i tried with direct links, next button was disabled. I wanted to play the next or previous song using default buttons(ids) 
**Note**
`android:id="@+id/exo_next` and `android:id="@+id/exo_prev` are placed in ` app:controller_layout_id="@layout/mediacontroller"`
Now I'm going to try your code with my app, I will let you know if I got an error.

**Comment by isuru18113** [comment]
@marcbaechinger Thank you very much. you are a lifesaver. It is working better than I expected. I have been trying to solve this for 2 days, but I could not do that. I have one more question with this,
I have set `app:use_artwork="false"` because I need to use custom artwork for songs. So how can I set to change the music image(musicimage) and music name(musicname) according to the song? since ArrayList handle by the exoplayer.

**Comment by isuru18113** [comment]
Hello, I'm back again. I finally found a way to change my UI according to the song. It is really easy. [https://exoplayer.dev/playlists.html](url) As this site said, we have to set a media tag for the for each audio and UI can be updated using the `public void onPositionDiscontinuity(@Player.DiscontinuityReason int reason)`. Finally, I solved my issue. It was very easy to use. Thank you, Exoplayer team 👍
